utility_analysis_new/histograms.py

```diff
@@ -319,24 +344,169 @@ def compute_dataset_histograms(col, data_extractors: pipeline_dp.DataExtractors,
+        partition_count_histogram, partition_privacy_id_count_histogram
+    ], backend)
 
-    histograms = backend.to_list(histograms, "Histograms to List")
-    # 1 element collection: [ContributionHistogram]
-    return backend.map(histograms, _list_to_contribution_histograms,
-                       "To ContributionHistograms")
-    # 1 element collection: (DatasetHistograms)
+
+########## Computing histograms on pre-aggregated datasets ####################
+# More details on pre-aggregate datatests are in docstring of function
+# pre_aggregation.preaggregate.
+
+
+def _compute_l0_contributions_histogram_on_preaggregted(
+        col, backend: pipeline_backend.PipelineBackend):
+    """Computes histogram of cross partition privacy id contributions.
+
+    This histogram contains: number of privacy ids which contributes to 1
+    partition, to 2 partitions etc.
+
+    Args:
+      col: collection with a pre-aggregated dataset, each element is
+      (partition_key, (count, sum, n_partitions)).
+      backend: PipelineBackend to run operations on the collection.
+
+    Returns:
+      1 element collection, which contains the computed Histogram.
+    """
+    col = backend.map_tuple(
+        col,
+        lambda _, x: x[2],  # x is (count, sum, n_partitions)
+        "Extract n_partitions")
+    # col: (int,), where each element is the number of partition the
+    # corresponding privacy_id contributes.
+    return _compute_frequency_histogram(col,
+                                        backend,
+                                        HistogramType.L0_CONTRIBUTIONS,
+                                        normalize=True)
+
+
+def _compute_linf_contributions_histogram_on_preaggregted(
+        col, backend: pipeline_backend.PipelineBackend):
+    """Computes histogram of per partition privacy id contributions.
+
+    This histogram contains: number of tuple (privacy id, partition_key) which
+    have 1 row in datasets, 2 rows etc.
+
+    Args:
+      col: collection with a pre-aggregated dataset, each element is
+      (partition_key, (count, sum, n_partitions)).
+      backend: PipelineBackend to run operations on the collection.
+
+    Returns:
+      1 element collection, which contains the computed Histogram.
+    """
+    linf = backend.map_tuple(
+        col,
+        lambda _, x: x[0],  # x is (count, sum, n_partitions)
+        "Extract per partition contribution")
+    # linf: (int,) where each element is the count of elements the
+    # corresponding privacy_id contributes to the partition.
+    return _compute_frequency_histogram(linf, backend,
+                                        HistogramType.LINF_CONTRIBUTIONS)
+
+
+def _compute_partition_count_histogram_on_preaggregted(
+        col, backend: pipeline_backend.PipelineBackend):
+    """Computes histogram of counts per partition.
+
+    This histogram contains: the number of partitions with count=1, with
+    count=2 etc.
```

Comment:
nit/opt: "This histogram contains: the number of partitions with total count of contributions = 1, 2"
to make it more explicit what "count" means in this context.

Reply:
thanks